### DISCUSSION TITLE
Translate the program

### DISCUSSION THREAD

**Discussion by imnotlxy** [helix-editor/helix#9161] [Ideas]
Nowadays programmers are all able to read and use English. However, translated UI can make them feel at home and boost their productivity.

Several frameworks are ready for implementing i18n. They are:
1. GNU `gettext`, wrapped by `gettextrs`.
2. Mozilla `fluent`.

With the help of `i18n-embed` crate, translation files of `fluent` can be embedded in the executable.

**Comment by imnotlxy** [comment]
COSMIC Settings can be an example using `fluent` and `i18n-embed`.


https://github.com/pop-os/cosmic-settings

**Reply by kirawi** [reply]
IMO `fluent` is pretty dependency-heavy and feels overkill for Helix. If this is a feature the maintainers want, then maybe something from scratch or simpler would be preferred. I can't speak on their behalf though.

**Reply by pascalkuthe** [reply]
I don't think a translation should be included in core. It takes a lot of effort would eat up revive bandwidth (or rather how do you review at all) and I don't see a ton of value in a translation. It would also make support much harder since we would get bug reports of people that use translated versions that we don't understand

**Reply by imnotlxy** [reply]
> I don't see a ton of value in a translation

Translated help messages and completion hints will be easier to read. Vim is translated.

> It would also make support much harder since we would get bug reports of people that use translated versions that we don't understand

Translate the program is a kind of virtue, while asking for support with screen-shots of those translated strings provided is rather rude. Related rules can be added to README (like using `LANG=en_US hx` command for reports), and you can simply refuse those who don't follow this.

> It takes a lot of effort would eat up revive bandwidth (or rather how do you review at all)

Sorry but I don't quite understand this. Take `fluent` as an example - string literals are replaced by message ids. Maybe having ids and contents in different files makes you feel disturbed?

> I don't think a translation should be included in core.

Can this be an extension instead?

> If this is a feature the maintainers want, then maybe something from scratch or simpler would be preferred.

Maybe we can add a `Messages` trait that includes message formatting functions? When a new translation starts, take Japanese, the translator should create a `Japanese` struct and implement the trait for it.

**Comment by gabrieleiro** [comment]
I'd love to contribute to translations for Portuguese if this gets implemented :)
